### PR TITLE
Module 6 - Data persistence

### PR DESCRIPTION
**Pull request by Juani-Sampedro**
-Information received from the API Rest calls saved into a Floor database.
-Provider implemented.

https://github.com/Juani-Sampedro/TheMovieDBApp/assets/84482037/be26845f-d764-48df-b6d8-9797b8106c66

-App name set natively and favIcon added.

image

### CODE REVIEW

**Comment by matiasilveiro-gl** [comment]
It'd be great to have a video of the app working!

File: lib/main.dart [hunk 1]
```diff
@@ -1,25 +1,49 @@
...
+class _MyAppState extends State<MyApp> {
+  late Initializer _initCore;
+  late AppRoutes _routes;
+
+  @override
+  void initState() {
+    super.initState();
+    _routes = AppRoutes();
+    _initCore = Initializer();
+    _initCore.initialize();
```
Comment (LeonardoMolleker) [inline lib/main.dart]: you are calling this method twice

File: lib/src/core/parameter/app_initializer.dart [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  MoviesBloc get upcomingMoviesBloc => _upcomingMoviesBloc;
+
+  MoviesBloc get popularMoviesBloc => _popularMoviesBloc;
+
+  GenresInMovieDetailBloc get genresBloc => _genresBloc;
+
+  FloorMovieDatabase get database => _floorDatabase;
+
+  Future<bool> initialize() async {
+    _floorDatabase = await $FloorFloorMovieDatabase.databaseBuilder('movie_database.db').build();
```
Comment (LeonardoMolleker) [inline lib/src/core/parameter/app_initializer.dart]: constant

File: lib/src/data/datasource/local/converters/list_int_converter.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  @override
+  List<int> decode(String databaseValue) {
+    final dynamic jsonFile = json.decode(databaseValue);
+    return List<int>.from(jsonFile['Genres']);
+  }
+
+  @override
+  String encode(List<int> value) {
+    final Map<String, dynamic> data = <String, dynamic>{};
+    data.addAll(<String, dynamic>{'Genres': value});
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/local/converters/list_int_converter.dart]: constant

File: lib/src/data/datasource/local/converters/list_string_converter.dart [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  @override
+  List<String?> decode(String databaseValue) {
+    var json = jsonDecode(databaseValue);
+    return List<String?>.from(json['categories']);
+  }
+
+  @override
+  String encode(List<String?> value) {
+    var map = <String, List<String?>>{};
+    map.addAll({'categories': value});
```
Comment (LeonardoMolleker) [inline lib/src/data/datasource/local/converters/list_string_converter.dart]: constant

File: lib/src/domain/usecase/implementation/genres_usecase_impl.dart [hunk 5]
```diff
@@ -1,16 +1,44 @@
...
 
 class GenresUseCase implements IUseCase {
   final GenreRepository repository;
+  final Connectivity _connectivity = Connectivity();
+  final GenreDatabase genreDatabase;
 
-  GenresUseCase(this.repository, {IApiService? apiService});
+  GenresUseCase({
+    required this.repository,
+    required this.genreDatabase,
```
Comment (LeonardoMolleker) [inline lib/src/domain/usecase/implementation/genres_usecase_impl.dart]: usecase shouldn't know the database directly, instead use a local repository as a intermediary

File: lib/main.dart (hunk 1 above)
Comment (Juani-Sampedro) [reply lib/main.dart]: solved

File: lib/src/core/parameter/app_initializer.dart (hunk 2 above)
Comment (Juani-Sampedro) [reply lib/src/core/parameter/app_initializer.dart]: solved

File: lib/src/data/datasource/local/converters/list_int_converter.dart (hunk 3 above)
Comment (Juani-Sampedro) [reply lib/src/data/datasource/local/converters/list_int_converter.dart]: solved

File: lib/src/data/datasource/local/converters/list_string_converter.dart (hunk 4 above)
Comment (Juani-Sampedro) [reply lib/src/data/datasource/local/converters/list_string_converter.dart]: solved

File: lib/src/domain/usecase/implementation/genres_usecase_impl.dart (hunk 5 above)
Comment (Juani-Sampedro) [reply lib/src/domain/usecase/implementation/genres_usecase_impl.dart]: solved

**Review by LeonardoMolleker — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

**Review by Kevin-Guillen — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]

**Review by matiasilveiro-gl — APPROVED** [review]

**Pull request merged**